src/platforms/esp32/components/avm_builtins/uart_driver.c

```diff
@@ -86,23 +95,34 @@ enum uart_cmd
     UARTInvalidCmd = 0,
     UARTReadCmd = 1,
     UARTWriteCmd = 2,
-    UARTCloseCmd = 3
+    UARTCloseCmd = 3,
+    UARTCancelCmd = 4
 };
 
 static const AtomStringIntPair cmd_table[] = {
     { ATOM_STR("\x4", "read"), UARTReadCmd },
     { ATOM_STR("\x5", "write"), UARTWriteCmd },
     { ATOM_STR("\x5", "close"), UARTCloseCmd },
+    { ATOM_STR("\xB", "cancel_read"), UARTCancelCmd },
     SELECT_INT_DEFAULT(UARTInvalidCmd)
 };
 
+static void safe_update_reader_data(struct UARTData *uart_data, term pid, uint64_t ref_ticks)
+{
+    SMP_MUTEX_LOCK(uart_data->reader_lock);
+    uart_data->reader_process_pid = pid;
+    uart_data->reader_ref_ticks = ref_ticks;
+    SMP_MUTEX_UNLOCK(uart_data->reader_lock);
+}
+
 EventListener *uart_interrupt_callback(GlobalContext *glb, EventListener *listener)
 {
     struct UARTData *uart_data = GET_LIST_ENTRY(listener, struct UARTData, listener);
 
     uart_event_t event;
     if (xQueueReceive(uart_data->rxqueue, (void *) &event, (TickType_t) portMAX_DELAY)) {
```

Comment:
Let's fix this: delay should be 0 here (this is what caused an issue with 5ac66d62f3fab4baa6d3568023e12eb72f6a20cb).

Indeed, listener is only called if data is available in the queue, so we never need to wait.

Reply:
By switching back to an if the issue that we were discussing on Telegram is no longer a problem, but I agree this can and should be 0 here. The problem was when I tried to make this a while loop, which doesn't work for the reason I posted about it as a follow-up to your comment about using `while` instead.